README
NAME
    MooseX::Blessed::Reconstruct - A Data::Visitor for creating Moose
    objects from blessed placeholders

VERSION
    version 0.5

SYNOPSIS
            use MooseX::Blessed::Reconstruct;


            my $obj = bless( {
                    init_arg_foo => "Blah",
                    arf => "yay",
            }, "Foo" );

            my $proper = MooseX::Blessed::Reconstruct->new->visit($obj);



            # equivalent to:

            my $proper = Foo->meta->new_object(%$obj);

            # but recursive (and works with shared references)

DESCRIPTION
    The purpose of this module is to "fix up" blessed data into a real Moose
    object.

    This is used internally by MooseX::YAML but has no implementation
    details having to do with YAML itself.

METHODS
    See Data::Visitor

    visit_object $object
        Calls "load_class" in Class::MOP on the "ref" of $object.

        If there's a metaclass, calls "visit_object_with_meta", otherwise
        "visit_ref" is used to walk the object brutishly.

        Returns a deep clone of the input structure with all the Moose
        objects reconstructed "properly".

    visit_object_with_meta $obj, $meta
        Uses the metaclass $meta to create a new instance, registers the
        instance with Data::Visitor's cycle tracking, and then inflates it
        using "new_object" in Moose::Meta::Class.

    prepare_args $obj
        Collapses $obj into key value pairs to be used as init args to
        "new_object" in Moose::Meta::Class.

AUTHORS
    *   Yuval Kogman <nothingmuch@woobling.org>

    *   Jonathan Rockway <jrockway@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2008 by Infinity Interactive, Yuval
    Kogman.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for MooseX-Blessed-Reconstruct

{{$NEXT}}
  [ API CHANGES ]

  [ BUG FIXES ]

  [ DOCUMENTATION ]

  [ ENHANCEMENTS ]

  [ NEW FEATURES ]

  [ STATISTICS ]

1.01 2018-09-26
  [ ENHANCEMENTS ]
    - Replace Load::Class by Module::Runtime. [GH#3, choroba]

  [ STATISTICS ]
    - code churn: 8 files changed, 199 insertions(+), 49 deletions(-)

1.00 2014-10-30
  [ BUG FIXES ]
    - $VERSION was still not using the dzil increment. (GH#1)

  [ STATISTICS ]
    - code churn: 3 files changed, 18 insertions(+), 12 deletions(-)

0.5 2014-10-30
  [ BUG FIXES ]
    - Class::MOP::load_class is deprecated, using Class::Load::load_class
      directly. (Yanick Champoux)

  [ STATISTICS ]
    - code churn: 7 files changed, 69 insertions(+), 52 deletions(-)

0.04 2010-07-13
  - Update BUILDALL behavior for recent versions of Moose (doy)

0.03 2009-09-17
  - Class::MOP no longer returns the metaclass from load_class

0.02 2008-12-06
  - Remove Class::Accessor from test suite (not having it as a dependency
    caused failures)

0.01 2008-09-24
  - Initial release


lib/MooseX/Blessed/Reconstruct.pm
package MooseX::Blessed::Reconstruct;
# ABSTRACT: A L<Data::Visitor> for creating Moose objects from blessed placeholders

use Moose 1.05;

use Carp qw(croak);

use Class::MOP 0.93;
use Module::Runtime;
use Data::Visitor 0.21; # n-arity visit

use Scalar::Util qw(reftype);

use namespace::clean -except => 'meta';

extends qw(Data::Visitor);

has load_classes => (
	isa => "Bool",
	is  => "rw",
	default => 1,
);

before visit_object => sub {
    my ( $v, $obj ) = @_;

    return unless $v->load_classes;

    Module::Runtime::use_package_optimistically(ref $obj);
};

sub visit_object {
	my ( $v, $obj ) = @_;

	my $meta = Class::MOP::get_metaclass_by_name(ref $obj);

    return ref $meta ? $v->visit_object_with_meta($obj, $meta)
                     : $v->visit_ref($obj);
}

sub visit_object_with_meta {
	my ( $v, $obj, $meta ) = @_;

	my $instance = $meta->get_meta_instance->create_instance;

	$v->_register_mapping( $obj => $instance );

	my $args = $v->prepare_args( $meta, $obj );

	$meta->new_object( %$args, __INSTANCE__ => $instance );

	return $instance;
}

my %refmap = (
    HASH   => sub { %{ $_[0] } },
    ARRAY  => sub { @{ $_[0] } },
    SCALAR => sub { ${ $_[0] } },
);

sub prepare_args {
	my ( $v, $meta, $obj ) = @_;

    my $f = $refmap{ reftype $obj } or croak "unknown ref type $obj";

    return $meta->name->BUILDARGS($v->visit($f->($obj)));
}

__PACKAGE__->meta->make_immutable;

__PACKAGE__

__END__

=pod

=head1 SYNOPSIS

	use MooseX::Blessed::Reconstruct;


	my $obj = bless( {
		init_arg_foo => "Blah",
		arf => "yay",
	}, "Foo" );

	my $proper = MooseX::Blessed::Reconstruct->new->visit($obj);


	# equivalent to:

	my $proper = Foo->meta->new_object(%$obj);

	# but recursive (and works with shared references)

=head1 DESCRIPTION

The purpose of this module is to "fix up" blessed data into a real Moose
object.

=head1 METHODS

See L<Data::Visitor>

=head2 new 

Constructor. 

=head3 arguments

=over 

=item load_classes 

If C<true> (which is the default), we will try to require its class 
when the target object is C<visit>ed. 

=back 

=head2 load_classes 

Read/write accessor to the C<load_classes> attribute.
If C<true>, we try to require its class when a target object is C<visit>ed. 

=head2 visit_object $object

Calls L<Class::MOP/get_metaclass_by_name($name)> on the C<ref> of $object.

If there's a metaclass, calls C<visit_object_with_meta>, otherwise C<visit_ref>
is used to walk the object brutishly.

Returns a deep clone of the input structure with all the L<Moose> objects
reconstructed "properly".

=head2 visit_object_with_meta $obj, $meta

Uses the metaclass C<$meta> to create a new instance, registers the instance
with L<Data::Visitor>'s cycle tracking, and then inflates it using
L<Moose::Meta::Class/new_object>.

=head2 prepare_args $obj

Collapses $obj into key value pairs to be used as init args to
L<Moose::Meta::Class/new_object>.



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "A L<Data::Visitor> for creating Moose objects from blessed placeholders",
  "dist_author" => [
    "Yuval Kogman <nothingmuch\@woobling.org>",
    "Jonathan Rockway <jrockway\@cpan.org>"
  ],
  "dist_name" => "MooseX-Blessed-Reconstruct",
  "dist_version" => "1.00",
  "license" => "perl",
  "module_name" => "MooseX::Blessed::Reconstruct",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Module::Runtime" => 0,
    "Class::MOP" => "0.93",
    "Data::Visitor" => "0.21",
    "Moose" => "1.05",
    "Scalar::Util" => 0,
    "Test::use::ok" => 0,
    "namespace::clean" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.88",
    "ok" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Test::More" => "0.88",
  "ok" => 0,
  "perl" => "5.006",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
